File: julia/bench_hhl.jl
using Pkg
Pkg.activate(joinpath(@__DIR__, "..", "julia_env"); io=devnull)

using Yao
using LinearAlgebra
using BenchmarkTools
using JSON
using Printf
using Statistics

function build_hhl_circuit()
    C = 2.0 / 3.0
    lambda1 = 2.0 / 3.0
    lambda2 = 4.0 / 3.0
    theta1 = 2.0 * asin(C / lambda1)
    theta2 = 2.0 * asin(C / lambda2)
    chain(4,
        put(2 => H), put(3 => H),
        put(2 => shift(3π / 2)), put(3 => shift(π / 2)),
        control(2, 4 => X), put(4 => shift(-3π / 4)),
        control(2, 4 => X), put(4 => shift(3π / 4)),
        put(3 => H), control(2, 3 => shift(-π / 2)),
        put(2 => H), swap(2, 3),
        control(2, 1 => Ry(theta1)), control(3, 1 => Ry(theta2)),
        swap(2, 3),
        put(2 => H), control(2, 3 => shift(π / 2)), put(3 => H),
        control(3, 4 => X), put(4 => shift(-π / 4)),
        control(3, 4 => X), put(4 => shift(π / 4)),
        control(2, 4 => X), put(4 => shift(-3π / 4)),
        control(2, 4 => X), put(4 => shift(3π / 4)),
        put(2 => H), put(3 => H),
    )
end

function run_prebuilt(circuit)
    reg = zero_state(4)
    reg |> circuit
    return statevec(reg)
end

function end_to_end()
    circuit = build_hhl_circuit()
    reg = zero_state(4)
    reg |> circuit
    return statevec(reg)
end

function simulate_measurements(circuit, nshots::Int)
    reg = zero_state(4)
    reg |> circuit
    return measure(reg; nshots=nshots)
end

function main()
    println("=== Ch.5 HHL 細粒度ベンチマーク (Julia / Yao.jl) ===")

    prebuilt = build_hhl_circuit()

    # warmup (JITコンパイルを起こす)
    run_prebuilt(prebuilt)
    end_to_end()
    simulate_measurements(prebuilt, 10)

    function summarize(bench)
        Dict(
            "samples" => length(bench.times),
            "min_ms" => minimum(bench.times) / 1e6,
            "median_ms" => median(bench.times) / 1e6,
            "mean_ms" => mean(bench.times) / 1e6,
            "max_ms" => maximum(bench.times) / 1e6,
            "stdev_ms" => std(bench.times) / 1e6,
        )
    end

    println("  -- build only --")
    b_build = @benchmark build_hhl_circuit() samples=200 evals=1
    display(b_build)
    println()

    println("  -- run only (prebuilt) --")
    b_run = @benchmark run_prebuilt($prebuilt) samples=200 evals=1
    display(b_run)
    println()

    println("  -- end-to-end (build + run) --")
    b_e2e = @benchmark end_to_end() samples=200 evals=1
    display(b_e2e)
    println()

    println("  -- simulate + measure 4096 shots --")
    b_measure = @benchmark simulate_measurements($prebuilt, 4096) samples=100 evals=1
    display(b_measure)
    println()

    out = Dict(
        "build_only" => summarize(b_build),
        "run_only" => summarize(b_run),
        "end_to_end" => summarize(b_e2e),
        "measure_4096" => summarize(b_measure),
    )

    @printf "\nSummary (ms):\n"
    for (k, v) in out
        @printf "  %-22s min=%.4f med=%.4f mean=%.4f\n" k v["min_ms"] v["median_ms"] v["mean_ms"]
    end

    out_path = joinpath(@__DIR__, "..", "results", "julia_bench.json")
    open(out_path, "w") do io
        JSON.print(io, out, 2)
    end
    println("Saved -> ", out_path)
end

main()

File: python/bench_hhl.py
"""
HHL 2x2 の細粒度ベンチマーク (Python / Qiskit)
- build のみ
- transpile のみ
- run (shots=1) のみ (transpile 済み回路を使う)
- run (statevector) のみ
- 端から端まで (build + transpile + run)

加えて、ランダムな 2x2 Hermitian 系を生成して解の質を統計する。
"""
import json
import time
import statistics as stats
from pathlib import Path

import numpy as np
from qiskit import QuantumCircuit, QuantumRegister, ClassicalRegister, transpile
from qiskit_aer import AerSimulator


def hhl_2x2_core(theta1, theta2):
    nb = QuantumRegister(1, 'b')
    nl = QuantumRegister(2, 'clock')
    na = QuantumRegister(1, 'ancilla')
    cr = ClassicalRegister(1, 'measure')
    qc = QuantumCircuit(na, nl, nb, cr)

    for qubit in nl:
        qc.h(qubit)
    qc.p(1.5 * np.pi, nl[0])
    qc.p(0.5 * np.pi, nl[1])
    qc.cx(nl[0], nb[0]); qc.p(-1.5 * np.pi / 2, nb[0])
    qc.cx(nl[0], nb[0]); qc.p(1.5 * np.pi / 2, nb[0])
    qc.cx(nl[1], nb[0]); qc.p(-0.5 * np.pi / 2, nb[0])
    qc.cx(nl[1], nb[0]); qc.p(0.5 * np.pi / 2, nb[0])
    qc.h(nl[1]); qc.cp(-np.pi / 2, nl[0], nl[1])
    qc.h(nl[0]); qc.swap(nl[0], nl[1])
    qc.cry(theta1, nl[0], na[0]); qc.cry(theta2, nl[1], na[0])
    qc.swap(nl[0], nl[1])
    qc.h(nl[0]); qc.cp(np.pi / 2, nl[0], nl[1]); qc.h(nl[1])
    qc.cx(nl[1], nb[0]); qc.p(-0.5 * np.pi / 2, nb[0])
    qc.cx(nl[1], nb[0]); qc.p(0.5 * np.pi / 2, nb[0])
    qc.cx(nl[0], nb[0]); qc.p(-1.5 * np.pi / 2, nb[0])
    qc.cx(nl[0], nb[0]); qc.p(1.5 * np.pi / 2, nb[0])
    for qubit in nl:
        qc.h(qubit)
    qc.measure(na[0], cr[0])
    return qc


def bench(fn, warmup=2, repeat=20):
    for _ in range(warmup):
        fn()
    ts = []
    for _ in range(repeat):
        t0 = time.perf_counter()
        fn()
        ts.append((time.perf_counter() - t0) * 1000)
    return dict(
        repeat=repeat, min_ms=min(ts), mean_ms=stats.mean(ts),
        median_ms=stats.median(ts), max_ms=max(ts),
        stdev_ms=stats.pstdev(ts),
    )


def main():
    sim = AerSimulator()
    C = 2 / 3
    theta1 = 2 * np.arcsin(C / (2 / 3))
    theta2 = 2 * np.arcsin(C / (4 / 3))

    qc_prebuilt = hhl_2x2_core(theta1, theta2)
    tqc_prebuilt = transpile(qc_prebuilt, sim)

    qc_sv = qc_prebuilt.remove_final_measurements(inplace=False)
    qc_sv.save_statevector()
    tqc_sv = transpile(qc_sv, sim)

    print("=== Ch.4 HHL 細粒度ベンチマーク (Python / Qiskit) ===")
    b_build = bench(lambda: hhl_2x2_core(theta1, theta2))
    print(f"  build-only            : {b_build}")
    b_transpile = bench(lambda: transpile(qc_prebuilt, sim))
    print(f"  transpile-only        : {b_transpile}")
    b_run_shots = bench(lambda: sim.run(tqc_prebuilt, shots=1).result().get_counts())
    print(f"  run (shots=1)         : {b_run_shots}")
    b_run_shots_4096 = bench(lambda: sim.run(tqc_prebuilt, shots=4096).result().get_counts())
    print(f"  run (shots=4096)      : {b_run_shots_4096}")
    b_run_sv = bench(lambda: sim.run(tqc_sv).result().get_statevector(qc_sv))
    print(f"  run (statevector)     : {b_run_sv}")
    b_end_to_end = bench(lambda: (
        lambda q: sim.run(transpile(q, sim), shots=4096).result().get_counts()
    )(hhl_2x2_core(theta1, theta2)))
    print(f"  end-to-end (shots=4096): {b_end_to_end}")

    # --- ランダム 2x2 Hermitian 系の解の質を統計
    print("\n=== ランダム 2x2 Hermitian 系 ===")
    rng = np.random.default_rng(42)
    n_trials = 10
    fids = []
    for trial in range(n_trials):
        # 対称にして実数固有値を保証
        M = rng.normal(size=(2, 2))
        A = (M + M.T) / 2
        A = A / np.linalg.norm(A) + np.eye(2)  # 正定値にシフト
        b = rng.normal(size=2)
        b = b / np.linalg.norm(b)
        x_cl = np.linalg.solve(A, b)
        x_cl_norm = x_cl / np.linalg.norm(x_cl)
        fids.append((trial, float(x_cl_norm[0]), float(x_cl_norm[1]), float(np.linalg.cond(A))))
    print(f"  {n_trials} 個のランダム系で古典解を計算 (HHL固定回路は A=[[1,-1/3],[-1/3,1]]のみなのでメタデータのみ)")
    for row in fids[:5]:
        print(f"    trial {row[0]}: x_norm=({row[1]:.3f}, {row[2]:.3f}), cond(A)={row[3]:.3f}")

    out = {
        "build_only": b_build,
        "transpile_only": b_transpile,
        "run_shots_1": b_run_shots,
        "run_shots_4096": b_run_shots_4096,
        "run_statevector": b_run_sv,
        "end_to_end": b_end_to_end,
        "random_systems_meta": fids,
    }
    p = Path(__file__).resolve().parent.parent / "results" / "python_bench.json"
    p.write_text(json.dumps(out, indent=2, ensure_ascii=False))
    print(f"\nSaved -> {p}")


if __name__ == "__main__":
    main()
